src/linear_algebra/adic_completion.lean

```diff
@@ -0,0 +1,189 @@
+
+/-- A module `M` is precomplete with respect to an ideal `I` if every Cauchy sequence converges. -/
+@[class] def is_precomplete : Prop :=
+∀ f : ℕ → M, (∀ {m n}, m ≤ n → f m ≡ f n [SMOD (I ^ m • ⊤ : submodule R M)]) →
+∃ L : M, ∀ n, f n ≡ L [SMOD (I ^ n • ⊤ : submodule R M)]
+
+/-- A module `M` is `I`-adically complete if it is Hausdorff and precomplete. -/
+@[class] def is_adic_complete : Prop :=
+is_Hausdorff I M ∧ is_precomplete I M
+
+/-- The Hausdorffification of a module with respect to an ideal. -/
+@[reducible] def Hausdorffification : Type* :=
+(⨅ n : ℕ, I ^ n • ⊤ : submodule R M).quotient
+
+/-- The completion of a module with respect to an ideal. This is not Hausdorff.
+In fact, this is only complete if the ideal is finitely generated. -/
+def adic_completion : submodule R (Π n : ℕ, (I ^ n • ⊤ : submodule R M).quotient) :=
+{ carrier := { f | ∀ {m n} (h : m ≤ n), liftq _ (mkq _)
+    (by { rw ker_mkq, exact smul_mono (ideal.pow_le_pow h) (le_refl _) }) (f n) = f m },
+  zero_mem' := λ m n hmn, by rw [pi.zero_apply, pi.zero_apply, linear_map.map_zero],
+  add_mem' := λ f g hf hg m n hmn, by rw [pi.add_apply, pi.add_apply,
+    linear_map.map_add, hf hmn, hg hmn],
+  smul_mem' := λ c f hf m n hmn, by rw [pi.smul_apply, pi.smul_apply, linear_map.map_smul, hf hmn] }
+
+namespace is_Hausdorff
+
+instance bot : is_Hausdorff (⊥ : ideal R) M :=
+λ x hx, by simpa only [pow_one ⊥, bot_smul, smodeq.bot] using hx 1
+
+variables {M}
+protected theorem subsingleton (h : is_Hausdorff (⊤ : ideal R) M) : subsingleton M :=
+⟨λ x y, eq_of_sub_eq_zero $ h (x - y) $ λ n, by { rw [ideal.top_pow, top_smul], exact smodeq.top }⟩
+variables (M)
+
+@[priority 100] instance of_subsingleton [subsingleton M] : is_Hausdorff I M :=
+λ x _, subsingleton.elim _ _
+
+variables {I M}
+theorem infi_pow_smul (h : is_Hausdorff I M) :
+  (⨅ n : ℕ, I ^ n • ⊤ : submodule R M) = ⊥ :=
+eq_bot_iff.2 $ λ x hx, (mem_bot _).2 $ h x $ λ n, smodeq.zero.2 $
+(mem_infi $ λ n : ℕ, I ^ n • ⊤).1 hx n
+
+end is_Hausdorff
+
+namespace Hausdorffification
+
+/-- The canonical linear map to the Hausdorffification. -/
+def of : M →ₗ[R] Hausdorffification I M := mkq _
+
+@[elab_as_eliminator]
+lemma induction_on {C : Hausdorffification I M → Prop} (x : Hausdorffification I M)
+  (ih : ∀ x, C (of I M x)) : C x :=
+quotient.induction_on' x ih
+
+instance : is_Hausdorff I (Hausdorffification I M) :=
+λ x, quotient.induction_on' x $ λ x hx, (quotient.mk_eq_zero _).2 $ (mem_infi _).2 $ λ n, begin
+  have := comap_map_mkq (⨅ n : ℕ, I ^ n • ⊤ : submodule R M) (I ^ n • ⊤),
+  simp only [sup_of_le_right (infi_le (λ n, (I ^ n • ⊤ : submodule R M)) n)] at this,
+  rw [← this, map_smul'', mem_comap, map_top, range_mkq, ← smodeq.zero], exact hx n
+end
+
+variables {M} [h : is_Hausdorff I N]
+include h
+
+/-- universal property of Hausdorffification: any linear map to a Hausdorff module extends to a
+unique map from the Hausdorffification. -/
```

Comment:
Probably worth explicitly stating the uniqueness?

Reply:
Just checking --- could you either add the uniqueness statement, or note in the comment that it's not actually being proved?

Otherwise, if you mark this as `request-review` again it's probably good to go soon.

Reply:
uniqueness proved